File: src/com/morph/engine/newgui/Button.java
package com.morph.engine.newgui;

import com.morph.engine.graphics.Texture;
import com.morph.engine.graphics.shaders.TransitionShader;
import com.morph.engine.math.Vector2f;
import com.morph.engine.util.RenderDataUtils;
import com.morph.engine.graphics.shaders.BasicTexturedShader;
import com.morph.engine.graphics.Color;
import com.morph.engine.physics.components.Transform2D;

public class Button extends Container {
    private String text;
	private String font;
	protected int size;

	private Texture hoverTexture;
	private Texture clickTexture;
	private Runnable onClick;

    public Button(String text, String font, int size, Color color, Color buttonColor, Texture texture, Texture altTexture, Transform2D transform, int depth) {
        super(transform,
				RenderDataUtils.createSquare(buttonColor, new TransitionShader(), texture, altTexture, 0),
				depth);
		this.text = text;
		this.font = font;
		this.size = size;

		this.hoverTexture = texture;
		this.clickTexture = texture;

		TextElement textObj = new TextElement(text, font, size, color,
				transform.getPosition().sub(transform.getScale().mul(new Vector2f(0.5f, 0.5f))), depth - 1);

		Vector2f textSize = textObj.getTopRight().sub(textObj.getBottomLeft()).abs();

		Vector2f shift = transform.getScale().sub(textSize).scale(0.5f);
		textObj.getTransform().translate(shift);

		this.addElement(textObj);
    }

	public void setOnClick(Runnable onClick) {
		this.onClick = onClick;
	}

	public String getText() {
		return text;
	}

	public String getFont() {
		return font;
	}
}

File: src/main/kotlin/com/morph/engine/newgui/Button.kt
package com.morph.engine.newgui

import com.morph.engine.graphics.Color
import com.morph.engine.graphics.LoadedFont
import com.morph.engine.graphics.Texture
import com.morph.engine.graphics.shaders.GUITintTransitionShader
import com.morph.engine.math.Vector2f
import com.morph.engine.physics.components.Transform2D
import com.morph.engine.util.RenderDataUtils

class Button(
        name: String,
        val text: String,
        val font: String,
        protected var size: Int,
        color: Color,
        buttonColor: Color,
        texture: Texture,
        altTexture: Texture,
        transform: Transform2D,
        depth: Int
) : Container(name, transform, RenderDataUtils.createTintedSquare(buttonColor, GUITintTransitionShader(), texture, altTexture, 0f), depth) {
    init {
        val textObj = TextElement(name + "-innerText", text, font, size,
                transform.position - (transform.scale * (Vector2f(0.5f, 0.5f))), color, depth - 1)

        val scaleRatio = size.toFloat() / LoadedFont.SIZE.toFloat()

        val xShift = (transform.scale.x - textObj.renderData.width * scaleRatio) * 0.5f
        val yShift = transform.scale.y * 0.5f - textObj.font.ascent * textObj.font.scale * 2.0f

        println(textObj.font.ascent * textObj.font.scale * scaleRatio)

        textObj.transform.translate(Vector2f(xShift, yShift))

        this.addElement(textObj)
    }
}
